plugins/pur/core/src/main/java/org/pentaho/di/repository/pur/PurRepository.java

```diff
@@ -407,14 +413,19 @@ public RepositoryDirectoryInterface createRepositoryDirectory( final RepositoryD
       if ( "/".equals( dir.getParent().getName() ) ) {
         throw new KettleException( BaseMessages.getString( PKG, "PurRepository.FailedDirectoryCreation.Message" ) );
       }
+
+      readWriteLock.writeLock().lock();
       RepositoryFile
-          newFolder =
-          pur.createFolder( dir.getParent().getObjectId() != null ? dir.getParent().getObjectId().getId() : null,
-              new RepositoryFile.Builder( dir.getName() ).folder( true ).build(), null );
+        newFolder =
+        pur.createFolder( dir.getParent().getObjectId() != null ? dir.getParent().getObjectId().getId() : null,
+          new RepositoryFile.Builder( dir.getName() ).folder( true ).build(), null );
+
       dir.setObjectId( new StringObjectId( newFolder.getId().toString() ) );
     } catch ( Exception e ) {
       throw new KettleException( "Unable to save repository directory with path [" + getPath( null, dir, null ) + "]",
-          e );
+        e );
+    } finally {
+      readWriteLock.writeLock().unlock();
```

Comment:
Could cause an IllegalMonitorStateException here. Your readWriteLock.writeLock().lock() is inside the try/finally block. So if you enter the method, throw the KettleException, then the finally is executed which attempts to release the lock you don't have yet. This needs to be reformulated as:
if ( "/".equals( dir.getParent().getName() ) {
  throw new KettleException ...
}
readWriteLock.writeLock().lock();
try {
   RepositoryFile newFolder = ... blah blah
   ... etc ...
} catch ( Exception e ) {
  throw new KettleException( ... blah blah );
} finally {
  readWriteLock.writeLock().unlock();
}